FrameLoadRequest and NavigationAction should share common members

nipunshukla21 opened the pull request:
#### 713fff3547d4e810d303839a8960328205ecfd1c
<pre>
FrameLoadRequest and NavigationAction should share common members
<a href="https://bugs.webkit.org/show_bug.cgi?id=298009">https://bugs.webkit.org/show_bug.cgi?id=298009</a>
<a href="https://rdar.apple.com/116202911">rdar://116202911</a>

Reviewed by Chris Dumez.

NavigationAction was often constructed using a FrameLoadRequest object, however because they
did not share common members, calling the NavigationAction ctor would often be clunky and
illegible. This change implements a new base class FrameLoadRequestBase containing shared
member variables and functions which both NavigationAction and FrameLoadRequest inherit from.
This change also adds a new constructor to NavigationAction which takes FrameLoadRequest&amp;
as an argument which is used where possible.

* Source/WebCore/loader/FrameLoadRequest.cpp:
* Source/WebCore/loader/FrameLoadRequest.h:
(WebCore::FrameLoadRequestBase::lockHistory const):
(WebCore::FrameLoadRequestBase::setLockHistory):
(WebCore::FrameLoadRequestBase::lockBackForwardList const):
(WebCore::FrameLoadRequestBase::setLockBackForwardList):
(WebCore::FrameLoadRequestBase::newFrameOpenerPolicy const):
(WebCore::FrameLoadRequestBase::setNewFrameOpenerPolicy):
(WebCore::FrameLoadRequestBase::shouldReplaceDocumentIfJavaScriptURL const):
(WebCore::FrameLoadRequestBase::setShouldReplaceDocumentIfJavaScriptURL):
(WebCore::FrameLoadRequestBase::shouldOpenExternalURLsPolicy const):
(WebCore::FrameLoadRequestBase::setShouldOpenExternalURLsPolicy):
(WebCore::FrameLoadRequestBase::downloadAttribute const):
(WebCore::FrameLoadRequestBase::sourceElement const):
(WebCore::FrameLoadRequestBase::setSourceElement):
(WebCore::FrameLoadRequestBase::initiatedByMainFrame const):
(WebCore::FrameLoadRequestBase::isRequestFromClientOrUserInput const):
(WebCore::FrameLoadRequestBase::setIsRequestFromClientOrUserInput):
(WebCore::FrameLoadRequestBase::isInitialFrameSrcLoad const):
(WebCore::FrameLoadRequestBase::setIsInitialFrameSrcLoad):
(WebCore::FrameLoadRequestBase::isContentRuleListRedirect const):
(WebCore::FrameLoadRequestBase::setIsContentRuleListRedirect):
(WebCore::FrameLoadRequestBase::isFromNavigationAPI const):
(WebCore::FrameLoadRequestBase::setIsFromNavigationAPI):
(WebCore::FrameLoadRequestBase::setDownloadAttribute):
(WebCore::FrameLoadRequestBase::setInitiatedByMainFrame):
(WebCore::FrameLoadRequest::disableShouldReplaceDocumentIfJavaScriptURL):
(WebCore::FrameLoadRequest::setIsRequestFromClientOrUserInput):
(WebCore::FrameLoadRequest::lockHistory const): Deleted.
(WebCore::FrameLoadRequest::setLockHistory): Deleted.
(WebCore::FrameLoadRequest::lockBackForwardList const): Deleted.
(WebCore::FrameLoadRequest::setLockBackForwardList): Deleted.
(WebCore::FrameLoadRequest::isInitialFrameSrcLoad const): Deleted.
(WebCore::FrameLoadRequest::setIsInitialFrameSrcLoad): Deleted.
(WebCore::FrameLoadRequest::isContentRuleListRedirect const): Deleted.
(WebCore::FrameLoadRequest::setIsContentRuleListRedirect): Deleted.
(WebCore::FrameLoadRequest::newFrameOpenerPolicy const): Deleted.
(WebCore::FrameLoadRequest::setNewFrameOpenerPolicy): Deleted.
(WebCore::FrameLoadRequest::shouldReplaceDocumentIfJavaScriptURL const): Deleted.
(WebCore::FrameLoadRequest::setShouldOpenExternalURLsPolicy): Deleted.
(WebCore::FrameLoadRequest::shouldOpenExternalURLsPolicy const): Deleted.
(WebCore::FrameLoadRequest::downloadAttribute const): Deleted.
(WebCore::FrameLoadRequest::sourceElement const): Deleted.
(WebCore::FrameLoadRequest::setSourceElement): Deleted.
(WebCore::FrameLoadRequest::initiatedByMainFrame const): Deleted.
(WebCore::FrameLoadRequest::isRequestFromClientOrUserInput const): Deleted.
(WebCore::FrameLoadRequest::isFromNavigationAPI const): Deleted.
(WebCore::FrameLoadRequest::setIsFromNavigationAPI): Deleted.
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::load):
(WebCore::createWindow):
* Source/WebCore/loader/NavigationAction.cpp:
(WebCore::NavigationAction::NavigationAction):
(WebCore::NavigationAction::copyWithShouldOpenExternalURLsPolicy const):
* Source/WebCore/loader/NavigationAction.h:
(WebCore::NavigationAction::shouldOpenExternalURLsPolicy const): Deleted.
(WebCore::NavigationAction::setShouldOpenExternalURLsPolicy): Deleted.
(WebCore::NavigationAction::initiatedByMainFrame const): Deleted.
(WebCore::NavigationAction::downloadAttribute const): Deleted.
(WebCore::NavigationAction::sourceElement const): Deleted.
(WebCore::NavigationAction::newFrameOpenerPolicy const): Deleted.
(WebCore::NavigationAction::setNewFrameOpenerPolicy): Deleted.
(WebCore::NavigationAction::lockHistory const): Deleted.
(WebCore::NavigationAction::setLockHistory): Deleted.
(WebCore::NavigationAction::lockBackForwardList const): Deleted.
(WebCore::NavigationAction::setLockBackForwardList): Deleted.
(WebCore::NavigationAction::shouldReplaceDocumentIfJavaScriptURL const): Deleted.
(WebCore::NavigationAction::setShouldReplaceDocumentIfJavaScriptURL):
(WebCore::NavigationAction::isRequestFromClientOrUserInput const): Deleted.
(WebCore::NavigationAction::setIsRequestFromClientOrUserInput): Deleted.
(WebCore::NavigationAction::isInitialFrameSrcLoad const): Deleted.
(WebCore::NavigationAction::setIsInitialFrameSrcLoad): Deleted.
(WebCore::NavigationAction::isContentRuleListRedirect const): Deleted.
(WebCore::NavigationAction::setIsContentRuleListRedirect): Deleted.
(WebCore::NavigationAction::isFromNavigationAPI const): Deleted.
(WebCore::NavigationAction::setIsFromNavigationAPI): Deleted.
* Source/WebKit/WebProcess/WebCoreSupport/WebRemoteFrameClient.cpp:
(WebKit::WebRemoteFrameClient::changeLocation):

Canonical link: <a href="https://commits.webkit.org/299706@main">https://commits.webkit.org/299706@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4b700fa0c684268f84cdca3665f822d5341fb1b0

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/119442 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/39129 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/29784 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/125712 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/71513 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/121319 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/39826 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/47710 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/90759 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/60064 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/122394 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/31784 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/107092 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/71236 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/30820 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/25201 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/69358 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/101241 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/25392 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/128687 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/46360 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/35091 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/99331 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/46725 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/103291 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/99143 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25268 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/44578 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/22586 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/42929 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/46223 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/51923 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/45688 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/49038 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/47375 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/loader/FrameLoadRequest.h:
```diff
@@ -40,7 +40,65 @@ class Element;
+    bool isFromNavigationAPI() const { return m_isFromNavigationAPI; }
+    void setIsFromNavigationAPI(bool isFromNavigationAPI) { m_isFromNavigationAPI = isFromNavigationAPI; }
+
+protected:
```
Protected data members is bad practice in C++ and discouraged in the WebKit codebase. Normally, those would be private with public (or protected) getters / setters.

cdumez commented on Source/WebCore/loader/FrameLoadRequest.h:
```diff
@@ -40,7 +40,65 @@ class Element;
-class FrameLoadRequest {
+class FrameLoadRequestBase {
+public:
+    FrameLoadRequestBase() = default;
```
I don't think we want a public constructor for this "base" class. Seems like the constructors should be protected.

cdumez replied:
Also, the destructor is not virtual so destroying a FrameLoadRequestBase object directly would be a bug. Right now, your destructor is implicit, public and non virtual, which is an issue. I recommend making the constructor protected and explicit (no need for virtual as I don't think we'll need polymorphism).

cdumez commented on Source/WebCore/loader/FrameLoadRequest.h:
```diff
@@ -40,7 +40,71 @@ class Element;
 
-class FrameLoadRequest {
+class FrameLoadRequestBase {
+protected:
```
Please move the protected members *below* the public ones.

cdumez commented on Source/WebCore/loader/NavigationAction.h:
```diff
@@ -121,43 +116,24 @@ class NavigationAction {
-    void setNewFrameOpenerPolicy(NewFrameOpenerPolicy newFrameOpenerPolicy) { m_newFrameOpenerPolicy = newFrameOpenerPolicy; }
+    // The shouldReplaceDocumentIfJavaScriptURL parameter will go away when the FIXME to eliminate the
+    // corresponding parameter from ScriptController::executeIfJavaScriptURL() is addressed.
+    void setShouldReplaceDocumentIfJavaScriptURL(ShouldReplaceDocumentIfJavaScriptURL shouldReplaceDocumentIfJavaScriptURL) { FrameLoadRequestBase::setShouldReplaceDocumentIfJavaScriptURL(shouldReplaceDocumentIfJavaScriptURL); }
```
Why don't we make this setter public in the base class instead of hiding the base class member function in the subclass?

nipunshukla21 replied:
That was an attempt from me to capture the original behavior. Previously, while the two classes duplicated the relevant member variable, `FrameLoadRequest` only had access to a disable function while `NavigationAction` had the full setter. Without protected member variables, I couldn't think of a simpler way of accomplishing this but I'd be happy to change it if there is one.

cdumez replied:
I don't think this is worth the extra code. I don't think it hurts much to make the corresponding setter public in the base class. And you can still have the convenience `disableShouldReplaceDocumentIfJavaScriptURL()` function in one of the subclasses which calls the parent setter.

cdumez commented on Source/WebCore/loader/NavigationAction.h:
```diff
@@ -30,6 +30,8 @@
 
 #include <WebCore/BackForwardItemIdentifier.h>
 #include <WebCore/CrossOriginOpenerPolicy.h>
+#include <WebCore/Element.h>
```
Do we really need this include?

nipunshukla21 replied:
Yeah, it seems like we do as when I remove it, I get a few incomplete type errors when we assign to `RefPtr<Element> m_sourceElement`.

cdumez replied:
That's happening in the base class though, in another header? Or did I misread?

cdumez replied:
I suspect the include is in the wrong header.

nipunshukla21 replied:
It's a bit odd that it works in the first place when including in this header but it definitely makes more sense to have the include in the base class's header so I've moved it there.

cdumez reviewed (approved):
LGTM if the bots are happy.

nipunshukla21 committed:
FrameLoadRequest and NavigationAction should share common members
https://bugs.webkit.org/show_bug.cgi?id=298009
rdar://116202911

Reviewed by Chris Dumez.

NavigationAction was often constructed using a FrameLoadRequest object, however because they
did not share common members, calling the NavigationAction ctor would often be clunky and
illegible. This change implements a new base class FrameLoadRequestBase containing shared
member variables and functions which both NavigationAction and FrameLoadRequest inherit from.
This change also adds a new constructor to NavigationAction which takes FrameLoadRequest&
as an argument which is used where possible.

* Source/WebCore/loader/FrameLoadRequest.cpp:
* Source/WebCore/loader/FrameLoadRequest.h:
(WebCore::FrameLoadRequestBase::lockHistory const):
(WebCore::FrameLoadRequestBase::setLockHistory):
(WebCore::FrameLoadRequestBase::lockBackForwardList const):
(WebCore::FrameLoadRequestBase::setLockBackForwardList):
(WebCore::FrameLoadRequestBase::newFrameOpenerPolicy const):
(WebCore::FrameLoadRequestBase::setNewFrameOpenerPolicy):
(WebCore::FrameLoadRequestBase::shouldReplaceDocumentIfJavaScriptURL const):
(WebCore::FrameLoadRequestBase::setShouldReplaceDocumentIfJavaScriptURL):
(WebCore::FrameLoadRequestBase::shouldOpenExternalURLsPolicy const):
(WebCore::FrameLoadRequestBase::setShouldOpenExternalURLsPolicy):
(WebCore::FrameLoadRequestBase::downloadAttribute const):
(WebCore::FrameLoadRequestBase::sourceElement const):
(WebCore::FrameLoadRequestBase::setSourceElement):
(WebCore::FrameLoadRequestBase::initiatedByMainFrame const):
(WebCore::FrameLoadRequestBase::isRequestFromClientOrUserInput const):
(WebCore::FrameLoadRequestBase::setIsRequestFromClientOrUserInput):
(WebCore::FrameLoadRequestBase::isInitialFrameSrcLoad const):
(WebCore::FrameLoadRequestBase::setIsInitialFrameSrcLoad):
(WebCore::FrameLoadRequestBase::isContentRuleListRedirect const):
(WebCore::FrameLoadRequestBase::setIsContentRuleListRedirect):
(WebCore::FrameLoadRequestBase::isFromNavigationAPI const):
(WebCore::FrameLoadRequestBase::setIsFromNavigationAPI):
(WebCore::FrameLoadRequestBase::setDownloadAttribute):
(WebCore::FrameLoadRequestBase::setInitiatedByMainFrame):
(WebCore::FrameLoadRequest::disableShouldReplaceDocumentIfJavaScriptURL):
(WebCore::FrameLoadRequest::setIsRequestFromClientOrUserInput):
(WebCore::FrameLoadRequest::lockHistory const): Deleted.
(WebCore::FrameLoadRequest::setLockHistory): Deleted.
(WebCore::FrameLoadRequest::lockBackForwardList const): Deleted.
(WebCore::FrameLoadRequest::setLockBackForwardList): Deleted.
(WebCore::FrameLoadRequest::isInitialFrameSrcLoad const): Deleted.
(WebCore::FrameLoadRequest::setIsInitialFrameSrcLoad): Deleted.
(WebCore::FrameLoadRequest::isContentRuleListRedirect const): Deleted.
(WebCore::FrameLoadRequest::setIsContentRuleListRedirect): Deleted.
(WebCore::FrameLoadRequest::newFrameOpenerPolicy const): Deleted.
(WebCore::FrameLoadRequest::setNewFrameOpenerPolicy): Deleted.
(WebCore::FrameLoadRequest::shouldReplaceDocumentIfJavaScriptURL const): Deleted.
(WebCore::FrameLoadRequest::setShouldOpenExternalURLsPolicy): Deleted.
(WebCore::FrameLoadRequest::shouldOpenExternalURLsPolicy const): Deleted.
(WebCore::FrameLoadRequest::downloadAttribute const): Deleted.
(WebCore::FrameLoadRequest::sourceElement const): Deleted.
(WebCore::FrameLoadRequest::setSourceElement): Deleted.
(WebCore::FrameLoadRequest::initiatedByMainFrame const): Deleted.
(WebCore::FrameLoadRequest::isRequestFromClientOrUserInput const): Deleted.
(WebCore::FrameLoadRequest::isFromNavigationAPI const): Deleted.
(WebCore::FrameLoadRequest::setIsFromNavigationAPI): Deleted.
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::load):
(WebCore::createWindow):
* Source/WebCore/loader/NavigationAction.cpp:
(WebCore::NavigationAction::NavigationAction):
(WebCore::NavigationAction::copyWithShouldOpenExternalURLsPolicy const):
* Source/WebCore/loader/NavigationAction.h:
(WebCore::NavigationAction::shouldOpenExternalURLsPolicy const): Deleted.
(WebCore::NavigationAction::setShouldOpenExternalURLsPolicy): Deleted.
(WebCore::NavigationAction::initiatedByMainFrame const): Deleted.
(WebCore::NavigationAction::downloadAttribute const): Deleted.
(WebCore::NavigationAction::sourceElement const): Deleted.
(WebCore::NavigationAction::newFrameOpenerPolicy const): Deleted.
(WebCore::NavigationAction::setNewFrameOpenerPolicy): Deleted.
(WebCore::NavigationAction::lockHistory const): Deleted.
(WebCore::NavigationAction::setLockHistory): Deleted.
(WebCore::NavigationAction::lockBackForwardList const): Deleted.
(WebCore::NavigationAction::setLockBackForwardList): Deleted.
(WebCore::NavigationAction::shouldReplaceDocumentIfJavaScriptURL const): Deleted.
(WebCore::NavigationAction::setShouldReplaceDocumentIfJavaScriptURL):
(WebCore::NavigationAction::isRequestFromClientOrUserInput const): Deleted.
(WebCore::NavigationAction::setIsRequestFromClientOrUserInput): Deleted.
(WebCore::NavigationAction::isInitialFrameSrcLoad const): Deleted.
(WebCore::NavigationAction::setIsInitialFrameSrcLoad): Deleted.
(WebCore::NavigationAction::isContentRuleListRedirect const): Deleted.
(WebCore::NavigationAction::setIsContentRuleListRedirect): Deleted.
(WebCore::NavigationAction::isFromNavigationAPI const): Deleted.
(WebCore::NavigationAction::setIsFromNavigationAPI): Deleted.
* Source/WebKit/WebProcess/WebCoreSupport/WebRemoteFrameClient.cpp:
(WebKit::WebRemoteFrameClient::changeLocation):

Canonical link: https://commits.webkit.org/299706@main
